File: 백준/Gold/2573. 빙산/빙산.py
from sys import stdin
input = stdin.readline
from collections import deque


def main():
    dx = [0, 0, 1, -1]
    dy = [1, -1, 0, 0]
    n, m = map(int, input().split())
    field = list(list(map(int, input().split())) for _ in range(n))
    glacier = deque()
    answer = 0

    def melt(glc: deque) -> deque:
        melted = []
        res = deque()

        while glc:
            y, x, t = glc.popleft()
            cnt = 0
            for k in range(4):
                ny, nx = y + dy[k], x + dx[k]
                if ny < 0 or nx < 0 or ny > n or nx > m: continue
                if field[ny][nx] == 0:
                    cnt += 1
            if field[y][x] - cnt > 0:
                res.append((y, x, field[y][x] - cnt))
                field[y][x] = field[y][x] - cnt
            else:
                melted.append((y, x))
        for i, j in melted:
            field[i][j] = 0
        return res

    def grouping(sy, sx):
        q = deque()
        q.append((sy, sx))
        visited[sy][sx] = True

        while q:
            y, x = q.popleft()
            for k in range(4):
                ny, nx = dy[k] + y, dx[k] + x
                if 1 <= ny < n-1 and 1 <= nx < m-1 and not visited[ny][nx] and field[ny][nx]:
                    visited[ny][nx] = True
                    q.append((ny, nx))

    for i in range(1, n-1):
        for j in range(1, m-1):
            if field[i][j]:
                glacier.append((i, j, field[i][j]))

    while glacier:
        answer += 1
        glacier = melt(glacier)
        # for i in range(n):
        #     print(*field[i])
        visited = [[False] * m for _ in range(n)]

        group = 0
        for i in range(1, n-1):
            for j in range(1, m-1):
                if field[i][j] and not visited[i][j]:
                    grouping(i, j)
                    group += 1
                if group > 1:
                    print(answer)
                    return
    print(0)


if __name__ == "__main__":
    main()

File: 백준/Gold/2573. 빙산/빙산.java
import java.io.*;
import java.util.*;

public class Main {
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int N, M;
		StringTokenizer st = new StringTokenizer(br.readLine());
		N = Integer.parseInt(st.nextToken());
		M = Integer.parseInt(st.nextToken());
		
		int[][] field = new int[N][M];
		// 컬렉션 어떻게 써야하는지 구조 알아보자 
		Queue<int[]> queue = new LinkedList<>();
		for (int i = 0; i < N; i++) {
			st = new StringTokenizer(br.readLine());
			for (int j = 0; j < M; j++) {
				field[i][j] = Integer.parseInt(st.nextToken());
				if (field[i][j] != 0) {
					queue.add(new int[] {i, j ,field[i][j]});
				}
			}
		}
		
		int year = 0;
		while (!queue.isEmpty()) {
			queue = melting(field, queue);
			year++;
			int group = 0;
			boolean[][] visited = new boolean[N][M];
			for (int i = 1; i < N - 1; i++) {
				for (int j = 1; j < M - 1; j++) {
					if (field[i][j] != 0 && !visited[i][j]) {
						grouping(field, visited, N, M, i, j);
						group++;
					}
					if (group > 1) {
						System.out.println(year);
						return;
					}
				}
			}
		}
		System.out.println(0);
	}
	
	// 왜 여기다 static 써야 되지?
	private static Queue<int[]> melting(int[][] field, Queue<int[]> glacier) {
		int[] dx = {1, -1, 0, 0};
		int[] dy = {0, 0, 1, -1};
		Queue<int[]> restGlacier = new LinkedList<>();
		List<int[]> melted = new ArrayList<>();
		
		while (!glacier.isEmpty()) {
			int[] data = glacier.poll();
			int y = data[0];
			int x = data[1];
			int h = data[2];
			
			int cnt = 0;
			for (int k = 0; k < 4; k++) {
				int nx = x + dx[k];
				int ny = y + dy[k];
				if (field[ny][nx] == 0) cnt++;
			}
			if (h - cnt > 0) {
				restGlacier.add(new int[] {y, x, h - cnt});
				field[y][x] = h - cnt;
			} else {
				melted.add(new int[] {y, x});
			}
		}
		for (int i = 0; i < melted.size(); i++) {
			field[melted.get(i)[0]][melted.get(i)[1]] = 0;
		}
		
		
		return restGlacier;
	}
	
	private static void grouping(int[][] field, boolean[][] visited, int n, int m, int startY, int startX) {
		int[] dx = {1, -1, 0, 0};
		int[] dy = {0, 0, 1, -1};
	
		Stack<int[]> stack = new Stack<>();
		stack.add(new int[]{startX, startY});
		visited[startY][startX] = true;
		while (!stack.isEmpty()) {
			int[] data = stack.pop();
			int x = data[0];
			int y = data[1];
			for (int k = 0; k < 4; k++) {
				int nx = x + dx[k];
				int ny = y + dy[k];
				
				if(nx < 1 || ny < 1 || nx >= m - 1 || ny >= n - 1 || field[ny][nx] == 0 || visited[ny][nx]) continue;
				stack.add(new int[] {nx, ny});
				visited[ny][nx] = true;
			}
		}
	}
}